quic/s2n-quic-transport/src/stream/controller/local_initiated.rs

```diff
@@ -283,3 +304,129 @@ impl ValueToFrameWriter<VarInt> for StreamsBlockedToFrameWriter {
+        Ok(())
+    }
+
+    #[inline]
+    fn on_packet_ack<A: ack::Set>(&mut self, ack_set: &A) {
+        let _ = ack_set;
+    }
+
+    #[inline]
+    fn on_packet_loss<A: ack::Set>(&mut self, ack_set: &A) {
+        let _ = ack_set;
+    }
+
+    #[inline]
+    fn close(&mut self) {}
+}
+
+/// Defines the open notify behavior for unidirectional streams
+///
+/// Since locally-initiated, unidirectional streams can only send data, all of
+/// these operations are no-op, as the peer will become aware of the stream once
+/// the local application starts sending on it.
+#[derive(Debug, Default)]
+pub(super) struct OpenNotifyUnidirectional;
+
+impl OpenNotifyBehavior for OpenNotifyUnidirectional {}
+
+impl transmission::interest::Provider for OpenNotifyUnidirectional {
+    #[inline]
+    fn transmission_interest<Q: transmission::interest::Query>(
+        &self,
+        _query: &mut Q,
+    ) -> transmission::interest::Result {
+        Ok(())
+    }
+}
+
+/// Defines the open notify behavior for bidirectional streams
+///
+/// This will send an empty STREAM frame to the peer for the largest
+/// opened, locally-initiated bidirectional stream. This prevents a deadlock
+/// where the local application opens the stream and tries to receive on it.
+#[derive(Debug, Default)]
+pub(super) struct OpenNotifyBidirectional {
+    max_opened: OnceSync<(), OpenNotifyFrameWriter>,
+}
+
+impl OpenNotifyBehavior for OpenNotifyBidirectional {
+    #[inline]
+    fn on_open_stream(&mut self) {
+        self.max_opened.force_delivery(())
+    }
+
+    #[inline]
+    fn on_transmit<W: WriteContext>(
+        &mut self,
+        stream_id: StreamId,
+        context: &mut W,
+    ) -> Result<(), OnTransmitError> {
+        self.max_opened.on_transmit(stream_id, context)
+    }
+
+    #[inline]
+    fn on_packet_ack<A: ack::Set>(&mut self, ack_set: &A) {
+        let _ = self.max_opened.on_packet_ack(ack_set);
+    }
+
+    #[inline]
+    fn on_packet_loss<A: ack::Set>(&mut self, ack_set: &A) {
+        self.max_opened.on_packet_loss(ack_set)
+    }
+
+    #[inline]
+    fn close(&mut self) {
+        self.max_opened.stop_sync();
+    }
+}
+
+impl transmission::interest::Provider for OpenNotifyBidirectional {
+    #[inline]
+    fn transmission_interest<Q: transmission::interest::Query>(
+        &self,
+        query: &mut Q,
+    ) -> transmission::interest::Result {
+        self.max_opened.transmission_interest(query)
+    }
+}
+
+/// Writes an empty STREAM frame to the peer, indicating the stream has been created
+#[derive(Debug, Default)]
+struct OpenNotifyFrameWriter;
+
+impl ValueToFrameWriter<()> for OpenNotifyFrameWriter {
+    #[inline]
+    fn write_value_as_frame<W: WriteContext>(
+        &self,
+        _value: (),
+        stream_id: StreamId,
+        context: &mut W,
+    ) -> Option<PacketNumber> {
+        context.write_frame(&frame::Stream {
+            stream_id: stream_id.into(),
+            is_last_frame: false,
+            is_fin: false,
```

Comment:
What happens if the application opens the stream and then immediately closes it? Do we send two empty stream frames?

Reply:
We would in that case, yeah.

Before, I had it where the `SendStream` was in charge of sending this if it didn't have any data to send and was essentially racing the application to fill it. If the application did send on it, then we would cancel the empty transmission. The downside to that was we would be potentially sending an empty frame per stream, rather than a frame per max stream like we do now, which would result in even more waste.